### PR TITLE
Improving debug on linux

### PR DESCRIPTION
**Pull request by leberus**
While fixing the thread's stuff on linux, I've realized that the way was written linux_debug.c would add some difficulties, so I decided to rewrite some parts. (no offense).

It's not done yet, not 100%: There are still some issues I need to solve, but I wanted to do a PR (without merging) to see if people is fine with the way I'm taking.

So far, some of the things that are working are:

* Attach to process and its threads
* Put breakpoingt, continue one thread, continue all threads, etc.
* **di** now shows more useful information but still incomplete:
```
[0x7f742fd214db]> di
type=none
signal=none
signum=0
sigpid=0
addr=0x0
inbp=false
pid=8371
tid=8371
uid=1000
gid=1000
exe=/bin/ls
cmdline=/bin/ls
cwd=/home/pp
stopreason=0
```
* **dpt** does not show "0" any more:
before:
```
[0x7f9e78c0e190]> dpt
 * 19495 s (current)
 - 0 s thread_0
 - 0 s thread_1
 * 19495 s ls
```
after:
```
[0x7f30bd716190]> dpt
 * 8347 s ls
```
* Status of the thread is fixed: fill_pid_info function takes care of that:
```

char *ptr;
        ptr = strstr (info, "State:");
        if (ptr) {
                switch (*(ptr + 7)) {
                case 'R':
                        pid_info->status = R_DBG_PROC_RUN;
                        break;
                case 'S':
                        pid_info->status = R_DBG_PROC_SLEEP;
                        break;
                case 'T':
                case 't':
                        pid_info->status = R_DBG_PROC_STOP;
                        break;
                case 'Z':
                        pid_info->status = R_DBG_PROC_ZOMBIE;
                        break;
                case 'X':
                        pid_info->status = R_DBG_PROC_DEAD;
                        break;
                default:
                        pid_info->status = R_DBG_PROC_SLEEP;
                        break;
                }
        }
```

**TODO:**

* Possibility to Ctrl-C gets us back to r2 prompt (as gdb does)
* Whenever a new thread arises, join them to the list and attach to it
* Get PC and Syscall for a pid
* Implement better wait_pid for Linux
* Fix all issues (segfaults, memleaks etc)

As I said, this is not for merging yet. I'll continue working on these days.
I think i didn't forger anything.. :S

### CODE REVIEW

**Comment by Maijin** [comment]
Wouldn't `di` information be better stored in sdb? #justasking

File: libr/debug/p/native/linux/linux_debug.c [hunk 1]
```diff
@@ -193,52 +206,193 @@ bool linux_set_options (RDebug *dbg, int pid) {
...
+// Local copy for us. Here we keep all traced threads (including the main_pid). 
+static RList *list_traced_threads = NULL;
+
+static int match_pid (const void *pid_o, const void *th_o) {
+	int pid = *(int *)pid_o;
+	RDebug *th = (RDebug *)th_o; 
+	return pid == th->pid;
+}
+
+static void remove_list_thread (void) {
```
Comment (alvarofe) [inline libr/debug/p/native/linux/linux_debug.c]: Within `remove_list_thread` just use `r_list_free` and set the free function in the list when creating it with `r_list_newf`

File: libr/debug/p/native/linux/linux_debug.c [hunk 2]
```diff
@@ -193,52 +206,193 @@ bool linux_set_options (RDebug *dbg, int pid) {
...
+	r_list_append (list_traced_threads, r_debug_pid_new (NULL, tid, 's', 0));
+}
+
+static int attach_to_ptid (RDebug *dbg, int ptid) {
+	linux_set_options (dbg, ptid);
+	eprintf ("attach_to_ptid: Attaching to pid: %d\n", ptid);
+	return ptrace (PTRACE_ATTACH, ptid, NULL, NULL);
+}
+
+static RList *attach_to_pid_and_threads (RDebug *dbg, int main_pid) {
```
Comment (alvarofe) [inline libr/debug/p/native/linux/linux_debug.c]: When declaring the function and the definition do not use spaces between the name and the parentheses, so it should be `attach_to_pid_and_threads(RDebug *dbg..)`

File: libr/debug/p/native/linux/linux_debug.c [hunk 3]
```diff
@@ -193,78 +202,285 @@ bool linux_set_options (RDebug *dbg, int pid) {
...
+	do {
+		int status;
+		int ret = waitpid (pid, &status, __WALL|WNOHANG);
+		eprintf ("ret: %d\n");
+		if (ret) {
+			reason = linux_ptrace_event (dbg, pid, status);
+
+			if (reason != R_DEBUG_REASON_UNKNOWN) {
+				break;
+			} else {
```
Comment (radare) [inline libr/debug/p/native/linux/linux_debug.c]: dont else after a break

File: libr/debug/p/native/linux/linux_debug.c [hunk 4]
```diff
@@ -193,78 +202,285 @@ bool linux_set_options (RDebug *dbg, int pid) {
...
+					reason = R_DEBUG_REASON_DEAD;
+				} else if (WIFSIGNALED (status)) {
+					eprintf ("child received signal %d\n", WTERMSIG (status));
+					reason = R_DEBUG_REASON_SIGNAL;
+				} else if (WIFSTOPPED (status)) {
+					if (WSTOPSIG (status) != SIGTRAP &&
+						WSTOPSIG (status) != SIGSTOP) {
+						eprintf ("child stopped with signal %d\n", WSTOPSIG (status));
+					}
+
```
Comment (radare) [inline libr/debug/p/native/linux/linux_debug.c]: empty line

File: libr/debug/p/native/linux/linux_debug.c [hunk 5]
```diff
@@ -193,78 +202,285 @@ bool linux_set_options (RDebug *dbg, int pid) {
...
+	return pid == th->pid;
+}
+
+static void add_and_attach_new_thread(RDebug *dbg, int tid) {
+	(void) linux_attach (dbg, tid);
+	eprintf ("add_and_attach_new_thread: buf: %d\n", tid);	
+	r_list_append (dbg->threads, r_debug_pid_new (NULL, tid, 's', 0));
+}
+
+static int attach_to_ptid(RDebug *dbg, int ptid) {
```
Comment (radare) [inline libr/debug/p/native/linux/linux_debug.c]: why not `attach_to_pid`?

File: libr/debug/p/native/linux/linux_debug.c [hunk 6]
```diff
@@ -193,78 +202,285 @@ bool linux_set_options (RDebug *dbg, int pid) {
...
+		}
+	}
+out:
 	return pid;
 }
 
-RDebugInfo *linux_info (RDebug *dbg, const char *arg) {
-	char procpid_cmdline[1024];
+static char *read_link(int pid, const char *file) {
+	char path[1024];
```
Comment (radare) [inline libr/debug/p/native/linux/linux_debug.c]: initialize with = {0};

File: libr/debug/p/native/linux/linux_debug.c [hunk 7]
```diff
@@ -193,78 +202,285 @@ bool linux_set_options (RDebug *dbg, int pid) {
...
+	char buf[1024];
```
Comment (radare) [inline libr/debug/p/native/linux/linux_debug.c]: use `MAX_PATH`

File: libr/debug/p/native/linux/linux_debug.c [hunk 8]
```diff
@@ -193,78 +202,285 @@ bool linux_set_options (RDebug *dbg, int pid) {
...
 	return rdi;
 }
 
-RList *linux_thread_list (int pid, RList *list) {
+static RDebugPid *fill_pid_info(const char *info, const char *path, int tid) {
+	RDebugPid *pid_info = R_NEW0 (RDebugPid);
+	if (!pid_info) {
+		return NULL;
+	} 
+	char *ptr;
```
Comment (radare) [inline libr/debug/p/native/linux/linux_debug.c]: define and assign in the same line

File: libr/debug/p/native/linux/linux_debug.c [hunk 9]
```diff
@@ -193,78 +202,285 @@ bool linux_set_options (RDebug *dbg, int pid) {
...
 	int i, thid = 0;
 	char *ptr, buf[1024];
 
 	if (!pid) {
 		r_list_free (list);
 		return NULL;
 	}
-	r_list_append (list, r_debug_pid_new ("(current)", pid, 's', 0));
 
+	eprintf ("linux_thread_list\n");
```
Comment (radare) [inline libr/debug/p/native/linux/linux_debug.c]: debug printf

File: libr/debug/p/native/linux/linux_debug.c [hunk 10]
```diff
@@ -193,78 +202,285 @@ bool linux_set_options (RDebug *dbg, int pid) {
...
 	/* if this process has a task directory, use that */
 	snprintf (buf, sizeof(buf), "/proc/%d/task", pid);
```
Comment (radare) [inline libr/debug/p/native/linux/linux_debug.c]: missing space before (

**Comment by radare** [comment]
Looks good! any regression so far? Please rebase and squash after fixing the comments. This will make travis happy

**Comment by leberus** [comment]
Hi! So far no regressions show up, but I need to do more tests.
I've fixed the comments (locally) btw. After some more changes and tests I'll rebase and squash.

thanks ;)

File: libr/include/r_debug.h [hunk 11]
```diff
@@ -267,12 +268,11 @@ typedef struct r_debug_info_t {
 	int signum;
 	void * lib;
```
Comment (radare) [inline libr/include/r_debug.h]: remove space after the asterisks

**Comment by radare** [comment]
lgtm, are you done or not yet? :D considering that there are 20 days until the release .it's a good time to merge so bugs and fixes can happen faster because of more people using/complaining.

let me know

**Comment by leberus** [comment]
Hi! I performed some tests and so far is working. There is still room for improvement, but I'd rather like to have this merged now, so if a bug/regression is being reported we can fix it bit by bit.
Hopefully it's gonna work fine :D

**Pull request merged**